File: 22_专题/区间问题/UnionIntervals.py
# IntervalUnion/UnionInterval/IntervalGraphUnion

from collections import defaultdict
from typing import Callable, DefaultDict, List, Tuple


class UnionFindArray:
    """元素是0-n-1的并查集写法,不支持动态添加

    初始化的连通分量个数 为 n
    """

    __slots__ = ("n", "part", "_parent", "_rank")

    def __init__(self, n: int):
        self.n = n
        self.part = n
        self._parent = list(range(n))
        self._rank = [1] * n

    def find(self, x: int) -> int:
        while self._parent[x] != x:
            self._parent[x] = self._parent[self._parent[x]]
            x = self._parent[x]
        return x

    def union(self, x: int, y: int) -> bool:
        """按秩合并."""
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return False
        if self._rank[rootX] > self._rank[rootY]:
            rootX, rootY = rootY, rootX
        self._parent[rootX] = rootY
        self._rank[rootY] += self._rank[rootX]
        self.part -= 1
        return True

    def unionTo(self, child: int, parent: int) -> bool:
        """定向合并.将child的父节点设置为parent."""
        rootX = self.find(child)
        rootY = self.find(parent)
        if rootX == rootY:
            return False
        self._parent[rootX] = rootY
        self._rank[rootY] += self._rank[rootX]
        self.part -= 1
        return True

    def unionWithCallback(self, x: int, y: int, f: Callable[[int, int], None]) -> bool:
        """
        f: 合并后的回调函数, 入参为 (big, small)
        """
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return False
        if self._rank[rootX] > self._rank[rootY]:
            rootX, rootY = rootY, rootX
        self._parent[rootX] = rootY
        self._rank[rootY] += self._rank[rootX]
        self.part -= 1
        f(rootY, rootX)
        return True

    def isConnected(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def getGroups(self) -> DefaultDict[int, List[int]]:
        groups = defaultdict(list)
        for key in range(self.n):
            root = self.find(key)
            groups[root].append(key)
        return groups

    def getRoots(self) -> List[int]:
        return list(set(self.find(key) for key in self._parent))

    def getSize(self, x: int) -> int:
        return self._rank[self.find(x)]

    def __repr__(self) -> str:
        return "\n".join(f"{root}: {member}" for root, member in self.getGroups().items())

    def __len__(self) -> int:
        return self.part


def unionIntervals(intervals: List[Tuple[int, int]]) -> "UnionFindArray":
    """
    给定n个区间[start,end), 返回合并后的并查集.
    """
    n = len(intervals)
    order = sorted(range(n), key=lambda i: (intervals[i][0], -intervals[i][1]))

    uf = UnionFindArray(n)
    keep = []
    for j in order:
        if keep:
            i = keep[-1]
            startI, endI = intervals[i]
            startJ, endJ = intervals[j]
            if endJ <= endI and endJ - startJ < endI - startI:
                uf.union(i, j)
                continue
        keep.append(j)

    for k in range(len(keep) - 1):
        i, j = keep[k], keep[k + 1]
        startI, endI = intervals[i]
        startJ, endJ = intervals[j]
        if max(startI, startJ) < min(endI, endJ):
            uf.union(i, j)

    return uf


if __name__ == "__main__":
    intervals = [(1, 3), (2, 4), (5, 6)]
    uf = unionIntervals(intervals)
    print(uf)

File: 22_专题/区间问题/UnionIntervals.go
// IntervalUnion/UnionInterval/IntervalGraphUnion
// 区间合并/合并区间

package main

import (
	"fmt"
	"sort"
	"strings"
)

func main() {
	fmt.Println(UnionIntervals([][2]int{{1, 3}, {3, 4}, {5, 7}, {6, 8}}))
}

// 给定n个区间[start,end), 返回合并后的并查集.
func UnionIntervals(intervals [][2]int) *UnionFindArray {
	n := len(intervals)
	order := make([]int, n)
	for i := 0; i < n; i++ {
		order[i] = i
	}
	sort.Slice(order, func(i, j int) bool {
		a, b := intervals[order[i]], intervals[order[j]]
		if a[0] == b[0] {
			return a[1] > b[1]
		}
		return a[0] < b[0]
	})

	uf := NewUnionFindArray(n)
	keep := make([]int, 0, n)
	for _, j := range order {
		if len(keep) > 0 {
			i := keep[len(keep)-1]
			startI, endI := intervals[i][0], intervals[i][1]
			startJ, endJ := intervals[j][0], intervals[j][1]
			if endJ <= endI && endJ-startJ < endI-startI {
				uf.Union(i, j)
				continue
			}
		}
		keep = append(keep, j)
	}

	for k := 0; k < len(keep)-1; k++ {
		i, j := keep[k], keep[k+1]
		startI, endI := intervals[i][0], intervals[i][1]
		startJ, endJ := intervals[j][0], intervals[j][1]
		if max(startI, startJ) < min(endI, endJ) {
			uf.Union(i, j)
		}
	}

	return uf
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

type UnionFindArray struct {
	// 连通分量的个数
	Part int
	n    int
	data []int
}

func NewUnionFindArray(n int) *UnionFindArray {
	data := make([]int, n)
	for i := 0; i < n; i++ {
		data[i] = -1
	}
	return &UnionFindArray{
		Part: n,
		n:    n,
		data: data,
	}
}

// 按秩合并.
func (ufa *UnionFindArray) Union(key1, key2 int) bool {
	root1, root2 := ufa.Find(key1), ufa.Find(key2)
	if root1 == root2 {
		return false
	}
	if ufa.data[root1] > ufa.data[root2] {
		root1 ^= root2
		root2 ^= root1
		root1 ^= root2
	}
	ufa.data[root1] += ufa.data[root2]
	ufa.data[root2] = root1
	ufa.Part--
	return true
}

func (ufa *UnionFindArray) UnionWithCallback(key1, key2 int, cb func(big, small int)) bool {
	root1, root2 := ufa.Find(key1), ufa.Find(key2)
	if root1 == root2 {
		return false
	}
	if ufa.data[root1] > ufa.data[root2] {
		root1 ^= root2
		root2 ^= root1
		root1 ^= root2
	}
	ufa.data[root1] += ufa.data[root2]
	ufa.data[root2] = root1
	ufa.Part--
	if cb != nil {
		cb(root1, root2)
	}
	return true
}

func (ufa *UnionFindArray) Find(key int) int {
	if ufa.data[key] < 0 {
		return key
	}
	ufa.data[key] = ufa.Find(ufa.data[key])
	return ufa.data[key]
}

func (ufa *UnionFindArray) IsConnected(key1, key2 int) bool {
	return ufa.Find(key1) == ufa.Find(key2)
}

func (ufa *UnionFindArray) GetSize(key int) int {
	return -ufa.data[ufa.Find(key)]
}

func (ufa *UnionFindArray) GetGroups() map[int][]int {
	groups := make(map[int][]int)
	for i := 0; i < ufa.n; i++ {
		root := ufa.Find(i)
		groups[root] = append(groups[root], i)
	}
	return groups
}

func (ufa *UnionFindArray) String() string {
	sb := []string{"UnionFindArray:"}
	groups := ufa.GetGroups()
	keys := make([]int, 0, len(groups))
	for k := range groups {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	for _, root := range keys {
		member := groups[root]
		cur := fmt.Sprintf("%d: %v", root, member)
		sb = append(sb, cur)
	}
	sb = append(sb, fmt.Sprintf("Part: %d", ufa.Part))
	return strings.Join(sb, "\n")
}
